src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Enter.java

```diff
@@ -419,6 +428,54 @@ public void visitPackageDef(JCPackageDecl tree) {
             }
         };
 
+    // Restructure top level to be an top level anonymous class.
+    public static void constructAnonymousMainClass(JCCompilationUnit tree,
```

Comment:
Question: any reason as to why this is done here and not in the parser? Typically we don't want to do tree transformation at parse time, as that messes up clients that want to access the "non-desugared" tree (such as IDE) and expect some mapping between source to AST. But if you do the rewriting in Enter, not much changes, that is, clients such as IDEs would still see something that doesn't resemble the source.

Reply:
The source file name is not available until after parsing.

Reply:
I see that logic in JavaCompiler - I wonder if that's just the way it is, or if there's a deeper reason as to why the sourcefile is set on the toplevel unit *after* parsing (I don't think I can see any, in which case that might be changed if that makes the rest of the code simpler). @lahodaj what do you think?

Reply:
I believe we were discussing this some time ago, and there were some problems. I don't recall the exact details, but I'll try to look into this later.

Reply:
I've sketched this:
https://github.com/lahodaj/jdk/commit/efe55f7d354ed7bbf91077d058823d682db501b9

I don't have too strong opinion here, there might be more cleanup possible after a change like this, and might be a bit cleaner outside of the parser; but forces the parser to work with files which is somewhat less clean.

Reply:
This is basically what I did originally until I decided that I needed a distinct class name. `log.currentSourceFile()` is the magic in this code. I'm willing to move this way if we all agree.